### ISSUE TITLE
Web Viewer - Brainstorm - Concepts on New Interaction Approaches - 3

### ISSUE THREAD

**Issue by gusferguson** [openmicroscopy/design#36]
Part 3 of the Web Viewer Brainstorming - following on from: https://github.com/openmicroscopy/design/issues/34

As these mockups go on, I am starting to run into problems with so many viewer options.
For this first round of mockups this will be dealt with by just working with the display controls/selections necessary for that immediate context. 
How to present the options and choices will need to be addressed once it is decided which options will be implemented.

**Clustering**

- clustering of ROIs

Still unhappy with this whole concept - I still hold to “if you have a few ROIs you want to look at them - if you have hundreds/thousands you probably don’t"
I am still not convinced it will be of much use unless there is a way of adding more information to this - e.g. showing what the ROIs are/their context/what folders they are in.
At the moment I can’t come up with any brilliant ideas for this.
Perhaps need to get more details on potential workflows from @bramalingam and others

Shown is a digital pathology image with clustering shown with and without a grid to help demarcation
Personally prefer grid

![1 viewer-clustering-zoom-1-nogrid](https://cloud.githubusercontent.com/assets/1552703/15040664/4dec53f6-12ae-11e6-8c48-66c8d63242e5.png)

![2 viewer-clustering-zoom-2-nogrid](https://cloud.githubusercontent.com/assets/1552703/15040665/4ded8690-12ae-11e6-81dc-6cbabbdecfc6.png)

![3 viewer-clustering-zoom-1](https://cloud.githubusercontent.com/assets/1552703/15040666/4df27d3a-12ae-11e6-93d0-12c30dd8e613.png)

![4 viewer-clustering-zoom-2](https://cloud.githubusercontent.com/assets/1552703/15040663/4deaaeca-12ae-11e6-88af-c0041aac1226.png)

**Grid Display:**

- showing all/selected Z and T as a grid with ROIs displayed

e.g. could be used to see the movement of an object marked by an ROI over time
Note: in general would be a good idea to make Z and T scrollers disappear if there is only 1 Z or 1 T

![5 viewer-z-t-as-grid](https://cloud.githubusercontent.com/assets/1552703/15040646/3ae23f00-12ae-11e6-9074-9c68d6d19cbf.png)

![6 viewer-t-as-grid-rhp](https://cloud.githubusercontent.com/assets/1552703/15040648/3afe2242-12ae-11e6-9483-391f39092d60.png)

![7 viewer-t-as-grid](https://cloud.githubusercontent.com/assets/1552703/15040647/3afcd69e-12ae-11e6-8e6e-c82e0f6de485.png)

**Intensity Display**

- Pixel intensity on mouse-over

Appreciate this has performance implications

![8 viewer-pixel-intensity](https://cloud.githubusercontent.com/assets/1552703/15040859/9eb00e26-12af-11e6-82a1-ca092fd91c68.png)

Comments and feedback as usual please.

**Comment by gusferguson** [comment]
**Pixel Intensity Heatmap**

This functionality would be accessed through the Settings tab of RHP - probably best in a fairly generic  "Settings Options" accordion tab.
This could then be used for other options such as look-up tables.
Tab is fairly unstructured at present as more options being discussed.

![viewer-heatmap-off](https://cloud.githubusercontent.com/assets/1552703/15070257/01a86aac-137b-11e6-9d68-43c9ab2af253.png)

![viewer-heatmap-on](https://cloud.githubusercontent.com/assets/1552703/15070258/01c17bdc-137b-11e6-8afa-bc5b1b67ee34.png)

**Comment by will-moore** [comment]
@gusferguson I agree with your comments on clustering. I think this is possibly useful for cases where the ROIs/shapes are not evenly distributed over the image (e.g. when we use google maps clustering to show which areas of the world have highest OME usage) but don't know how applicable to imaging use case. Need to try this on real data. @sbesson I seem to remember you had nice point tracking data with many shapes per plane? This might be useful to import and try clustering? Also clustering might be useful just to improve JavaScript performance (not rendering many shapes) but with many thousands of shapes it's possible that we reach a bottleneck of getting data to the browser before we hit JavaScript rendering issues?
The grid display is something I've considered a bit before since it could be handy. You can see a prototype at /webtest/image_dimensions/<imageId>/ (screenshot below). However, I think you'd need to have each panel as a separate viewer (like in OMERO.figure) so that when you zoom in to see image detail, you can still see the full grid - Not like the current split view in web/insight.

![screen shot 2016-05-09 at 16 06 04](https://cloud.githubusercontent.com/assets/900055/15117572/194ff31e-1600-11e6-8ead-8671e2bf2b0a.png)

Another alternative to this is to offer a 'stack' or 'grid' of images in a lens (E.g. click "show lens" on demo at http://codepen.io/will-moore/pen/Beuyc)

![screen shot 2016-05-09 at 16 13 09](https://cloud.githubusercontent.com/assets/900055/15117786/0f5b4b5a-1601-11e6-8a47-41d2487ef1c6.png)

For lookup tables, this is really just another colour mapping for a channel, so when you click the 'colour-picker' for a channel, you can choose a range of lookup tables as well as single-colour options.
Try playing with Icy as an example. They also allow you to edit your lookup table, but I think this is probably an overkill for us.

**Comment by gusferguson** [comment]
Additional mockups showing histogram options:

1. Solid histogram

Histogram only shows data for the channel corresponding to active/selected slider
Max/min boundary lines same colour as for the active/selected slider channel
Max/min not adjustable on histogram - only via slider

![viewer-histogram-solid](https://cloud.githubusercontent.com/assets/1552703/15147335/1c4b02f8-16b8-11e6-84ef-72064bf7ab3f.jpg)

2. Line traces

"Histogram" shows traces for all channels
Only Max/min boundaries for the active/selected slider - and same colour as channel
Max/min not adjustable on histogram - only via slider

![viewer-histogram-lines](https://cloud.githubusercontent.com/assets/1552703/15147339/229e3026-16b8-11e6-8c04-c1ff895db5d7.jpg)

**Comment by pwalczysko** [comment]
re 
> For heat maps - need to discuss if facility to set thresholds/vary colours should be offered.

I think that the feature which is a downgraded version of both of these concenpts should be offered, as it was already at least once by users: Warning colors (red for too high, blue for too low) for pixels which are at 255 and 0 intensities respectively. This is a standard feature on confocal and other microscopy software. Probably a candidate for the Setttings tab.

**Comment by gusferguson** [comment]
**Include look-up-tables in colour picker drop-down options.**

Will probably be stuck with just text - ? if able to generate icon
No facility to add or edit LUTs in UI

![viewer-colour-chooser-lut-dd](https://cloud.githubusercontent.com/assets/1552703/15151299/7a0e9b74-16c8-11e6-961d-b4c60771dd9c.jpg)

**Comment by will-moore** [comment]
@pwalczysko yes, showing saturated as red and black as blue is useful and has been requested. But this is simply another lookup table to add to the list (see @gusferguson's last screenshot). Just need to work out a name for it!

**Comment by jburel** [comment]
@gusferguson: we might only go for text for now for LUT.
One thing we discussed this afternoon with @will-moore is the representation on the button when a LUT is used. We still have not found  a suitable solution.
